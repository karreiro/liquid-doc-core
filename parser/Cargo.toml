[package]
name = "liquid_doc_parser"
version = "0.1.0"
edition = "2021"

[lib]
crate-type = ["rlib"]

[dependencies]
serde.workspace = true
serde_json.workspace = true
pest = "2.7"
pest_derive = "2.7"

[dev-dependencies]
insta.workspace = true
paste = "1.0"
pretty_assertions.workspace = true
