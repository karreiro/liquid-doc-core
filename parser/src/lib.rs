mod ast;
mod parser;

// Re-export all AST types
pub use ast::*;

// Re-export parser functionality
pub use parser::{parse_liquid_string, visit, LiquidParser};
