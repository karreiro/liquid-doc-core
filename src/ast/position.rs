use serde::{Deserialize, Serialize};

use crate::parser::Rule;

#[derive(Serialize, Deserialize, Debug, Copy, Clone, PartialEq, Eq)]
pub struct Position {
    pub start: usize,
    pub end: usize,
}

impl Position {
    pub fn new(start: usize, end: usize, offset: Option<usize>) -> Self {
        match offset {
            Some(o) => Position {
                start: start + o,
                end: end + o,
            },
            None => Position { start, end },
        }
    }
    pub fn from_pair(pair: &pest::iterators::Pair<Rule>, offset: Option<usize>) -> Self {
        let span = pair.as_span();
        Position::new(span.start(), span.end(), offset)
    }

    pub fn shift_start(&mut self, offset: usize) {
        self.start += offset;
    }
}
