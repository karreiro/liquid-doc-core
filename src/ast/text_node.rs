use serde::{Deserialize, Serialize};

use crate::parser::Rule;

use super::position::Position;

#[derive(Serialize, Deserialize, Debug, Clone, PartialEq)]
pub struct TextNode {
    pub value: String,
    pub position: Position,
    pub source: String,
    #[serde(rename = "type", skip_deserializing)]
    pub type_: &'static str,
}

impl TextNode {
    pub fn new(value: String, position: Position, source: String) -> Self {
        TextNode {
            value,
            position,
            source,
            type_: "TextNode",
        }
    }
    pub fn from_pair(pair: &pest::iterators::Pair<Rule>, position_offset: Option<usize>) -> Self {
        let source_str = pair.as_str();
        TextNode::new(
            source_str.to_string(),
            Position::from_pair(pair, position_offset),
            source_str.to_string(),
        )
    }

    pub fn without_brackets(
        pair: &pest::iterators::Pair<Rule>,
        position_offset: Option<usize>,
    ) -> Self {
        let mut text_node = Self::from_pair(pair, position_offset);
        let new_value = text_node
            .value
            .trim_matches(|c| c == '{' || c == '}' || c == '[' || c == ']')
            .to_string();
        if new_value == text_node.value {
            return text_node; // No change needed
        }
        text_node.value = new_value;

        text_node.position.shift_start(1); // Adjust position to account for removed brackets
        text_node.position.shift_end_down(1); // Adjust end position as well
        text_node
    }

    pub fn trim_content_start(&mut self, to_strip: &str) {
        if self.value.starts_with(to_strip) {
            self.value = self.value.trim_start_matches(to_strip).to_string();
            self.position.shift_start(to_strip.len());
        }
    }

    #[allow(dead_code)]
    pub fn as_str(&self) -> &str {
        &self.value
    }
    pub fn is_empty(&self) -> bool {
        self.value.is_empty()
    }
}
