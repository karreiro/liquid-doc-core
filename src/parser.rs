use crate::ast::{LiquidDocParamNode, *};
use pest::Parser;
use pest_derive::Parser;
use wasm_bindgen::{prelude::wasm_bindgen, JsValue};

#[derive(Parser)]
#[grammar = "liquid.pest"]
pub struct LiquidParser;

pub fn visit(
    ast: &mut LiquidAST,
    pair: pest::iterators::Pair<Rule>,
    position_offset: Option<usize>,
) {
    match pair.as_rule() {
        Rule::Document => {
            // The Document rule is the root of the AST, so we can just ignore it
            for inner_pair in pair.into_inner() {
                visit(ast, inner_pair, position_offset);
            }
        }
        Rule::ImplicitDescription => {
            let node = LiquidDocDescriptionNode::implicit(&pair, position_offset);
            if !node.content.is_empty() {
                ast.add_node(LiquidNode::LiquidDocDescriptionNode(node));
            }
        }
        Rule::LiquidDocNode => {
            let mut content = pair.into_inner();
            let next = content.next().unwrap();
            match next.as_rule() {
                Rule::paramNode => {
                    let node = LiquidDocParamNode::new(&next, position_offset);

                    ast.add_node(LiquidNode::LiquidDocParamNode(node));
                }
                Rule::exampleNode => {
                    let node = LiquidDocExampleNode::from_pair(&next, position_offset);
                    ast.add_node(LiquidNode::LiquidDocExampleNode(node));
                }
                Rule::descriptionNode => {
                    let node = LiquidDocDescriptionNode::explicit(&next, position_offset);
                    ast.add_node(LiquidNode::LiquidDocDescriptionNode(node));
                }
                Rule::promptNode => {
                    // Process prompt node
                    unimplemented!("Prompt nodes are not yet implemented");
                }
                Rule::fallbackNode => {
                    // Process fallback node
                    unimplemented!("Fallback nodes are not yet implemented");
                }
                _ => unreachable!("Unexpected rule in LiquidDocNode: {:?}", next.as_rule()),
            }
        }
        Rule::TextNode => {
            let text_node = TextNode::from_pair(&pair, position_offset);
            if !text_node.is_empty() {
                ast.add_node(LiquidNode::TextNode(text_node));
            }
        }
        _ => unimplemented!("Handle rule: {:?}", pair.as_rule()),
    }
}

pub(crate) fn parse_liquid_string(
    input: &str,
    position_offset: Option<usize>,
) -> Option<LiquidAST> {
    let text = LiquidParser::parse(Rule::Document, input)
        .map_err(|e| println!("Parsing error: {}", e))
        .ok()?;

    let mut ast = LiquidAST::new();
    for pair in text {
        visit(&mut ast, pair, position_offset);
    }

    Some(ast)
}

#[wasm_bindgen]
pub fn parse_liquid(input: &str) -> JsValue {
    serde_wasm_bindgen::to_value(&parse_liquid_string(input, None))
        .expect("The LiquidAst was not in the correct format")
}
