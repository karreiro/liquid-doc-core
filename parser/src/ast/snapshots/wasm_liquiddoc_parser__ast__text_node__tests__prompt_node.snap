---
source: src/ast/text_node.rs
expression: "@prompt Make something pretty"
---
[
  {
    "type": "TextNode",
    "value": "@prompt Make something pretty",
    "position": {
      "start": 10,
      "end": 39
    },
    "source": "@prompt Make something pretty"
  }
]
